File: packages/zod-openapi/src/interface-generator.ts
/**
 * Generates TypeScript interface/type strings from OpenAPI schemas.
 *
 * This produces concrete types that appear directly in .d.ts files,
 * rather than requiring z.infer<> resolution at the type level.
 */

import {
  getSchemaExportedVariableNameForPrimitiveType,
  getSchemaExportedVariableNameForStringFormat,
} from "./registry";
import type { AnySchema } from "./types/types";

const validIdentifierRegex = /^[a-zA-Z_$][a-zA-Z0-9_$]*$/;

type SchemaToTypeOptions = {
  outputSchemaNames?: Set<string>;
};

function quotePropertyName(name: string): string {
  return validIdentifierRegex.test(name) ? name : `'${name}'`;
}

function toPascalCase(name: string): string {
  return name
    .replace(/([a-z0-9])([A-Z])/g, "$1 $2")
    .replace(/[^a-zA-Z0-9]/g, " ")
    .split(" ")
    .filter(Boolean)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join("");
}

export function schemaExportNameToOutputAlias(name: string): string {
  return `${toPascalCase(name)}Output`;
}

function registerOutputSchemaName(
  schemaName: string,
  options?: SchemaToTypeOptions,
): string {
  options?.outputSchemaNames?.add(schemaName);
  return schemaExportNameToOutputAlias(schemaName);
}

function getRegisteredOutputAlias(
  schema: AnySchema,
  options?: SchemaToTypeOptions,
): string | undefined {
  if (!schema || typeof schema !== "object") return undefined;

  if (schema["type"] === "string" && typeof schema["format"] === "string") {
    const customSchemaName = getSchemaExportedVariableNameForStringFormat(
      schema["format"],
    );
    if (customSchemaName) {
      return registerOutputSchemaName(customSchemaName, options);
    }
  }

  if (
    schema["type"] === "number" ||
    schema["type"] === "integer" ||
    schema["type"] === "boolean"
  ) {
    const customSchemaName = getSchemaExportedVariableNameForPrimitiveType(
      schema["type"],
    );
    if (customSchemaName) {
      return registerOutputSchemaName(customSchemaName, options);
    }
  }

  return undefined;
}

/**
 * Converts an OpenAPI schema to a TypeScript type string.
 *
 * @example
 * schemaToTypeString({ type: 'string' }) // => 'string'
 * schemaToTypeString({ type: 'object', properties: { id: { type: 'string' } } }) // => '{ id?: string }'
 */
export function schemaToTypeString(
  schema: AnySchema,
  options?: SchemaToTypeOptions,
): string {
  if (!schema || typeof schema !== "object") return "unknown";

  // Handle $ref
  if (schema["$ref"] && typeof schema["$ref"] === "string") {
    const match = (schema["$ref"] as string).match(
      /#\/components\/schemas\/(.+)/,
    );
    let result = "unknown";
    if (match && match[1]) {
      // Decode URI-encoded schema names (e.g., %20 -> space)
      result = decodeURIComponent(match[1]);
    }
    if (schema["nullable"] === true) {
      result = `(${result} | null)`;
    }
    return result;
  }

  let result: string = "unknown";

  // Handle oneOf (union)
  if ("oneOf" in schema && Array.isArray(schema["oneOf"])) {
    const unionMembers = (schema["oneOf"] as AnySchema[]).map((s) =>
      schemaToTypeString(s, options),
    );
    result = unionMembers.length > 1 ? `(${unionMembers.join(" | ")})` : unionMembers[0] ?? "unknown";
  }
  // Handle allOf (intersection)
  else if ("allOf" in schema && Array.isArray(schema["allOf"])) {
    const intersectionMembers = (schema["allOf"] as AnySchema[]).map((s) =>
      schemaToTypeString(s, options),
    );
    result = intersectionMembers.length > 1
      ? `(${intersectionMembers.join(" & ")})`
      : intersectionMembers[0] ?? "unknown";
  }
  // Handle anyOf (union, similar to oneOf)
  else if ("anyOf" in schema && Array.isArray(schema["anyOf"])) {
    const unionMembers = (schema["anyOf"] as AnySchema[]).map((s) =>
      schemaToTypeString(s, options),
    );
    result = unionMembers.length > 1 ? `(${unionMembers.join(" | ")})` : unionMembers[0] ?? "unknown";
  }
  // Handle type-based schemas
  else {
    switch (schema["type"]) {
      case "string": {
        const registeredAlias = getRegisteredOutputAlias(schema, options);
        if (registeredAlias) {
          result = registeredAlias;
        } else if (schema["enum"] && Array.isArray(schema["enum"])) {
          // String enum
          result = (schema["enum"] as string[])
            .map((v) => JSON.stringify(v))
            .join(" | ");
        } else {
          result = "string";
        }
        break;
      }
      case "number":
      case "integer": {
        const registeredAlias = getRegisteredOutputAlias(schema, options);
        if (registeredAlias) {
          result = registeredAlias;
        } else if (schema["enum"] && Array.isArray(schema["enum"])) {
          // Numeric enum
          result = (schema["enum"] as number[]).map((v) => String(v)).join(" | ");
        } else {
          result = "number";
        }
        break;
      }
      case "boolean":
        result = getRegisteredOutputAlias(schema, options) ?? "boolean";
        break;
      case "null":
        result = "null";
        break;
      case "array":
        if (schema["items"]) {
          const itemType = schemaToTypeString(schema["items"] as AnySchema, options);
          result = `Array<${itemType}>`;
        } else {
          result = "unknown[]";
        }
        break;
      case "object":
        result = objectSchemaToTypeString(schema, options);
        break;
      default:
        // Try to detect object from properties
        if (schema["properties"]) {
          result = objectSchemaToTypeString(schema, options);
        } else if (schema["enum"] && Array.isArray(schema["enum"])) {
          // Untyped enum
          result = (schema["enum"] as unknown[])
            .map((v) => JSON.stringify(v))
            .join(" | ");
        } else {
          result = "unknown";
        }
        break;
    }
  }

  // Handle nullable
  if (schema["nullable"] === true) {
    result = `(${result} | null)`;
  }

  return result;
}

/**
 * Converts an OpenAPI object schema to a TypeScript object type string.
 */
function objectSchemaToTypeString(
  schema: AnySchema,
  options?: SchemaToTypeOptions,
): string {
  const properties = schema["properties"] as Record<string, AnySchema> | undefined;
  const required = new Set((schema["required"] as string[]) ?? []);
  const additionalProperties = schema["additionalProperties"];

  if (!properties && !additionalProperties) {
    return "Record<string, unknown>";
  }

  const propertyStrings: string[] = [];

  if (properties) {
    for (const [propName, propSchema] of Object.entries(properties)) {
      const isRequired = required.has(propName);
      const propType = schemaToTypeString(propSchema, options);
      const quotedName = quotePropertyName(propName);
      propertyStrings.push(
        `${quotedName}${isRequired ? "" : "?"}: ${propType}`,
      );
    }
  }

  // Handle additionalProperties
  if (additionalProperties === true) {
    propertyStrings.push("[key: string]: unknown");
  } else if (
    typeof additionalProperties === "object" &&
    additionalProperties !== null
  ) {
    const additionalType = schemaToTypeString(additionalProperties as AnySchema, options);
    propertyStrings.push(`[key: string]: ${additionalType}`);
  }

  return `{ ${propertyStrings.join("; ")} }`;
}

/**
 * Generates a full TypeScript interface declaration.
 *
 * @example
 * generateInterface('User', { type: 'object', properties: { id: { type: 'string' } }, required: ['id'] })
 * // => 'export interface User { id: string; }'
 */
export function generateInterface(
  name: string,
  schema: AnySchema,
  options?: SchemaToTypeOptions,
): string {
  const properties = schema["properties"] as Record<string, AnySchema> | undefined;
  const required = new Set((schema["required"] as string[]) ?? []);

  // For non-object types, use type alias instead of interface
  if (schema["type"] !== "object" && !properties) {
    return `export type ${name} = ${schemaToTypeString(schema, options)};`;
  }

  const lines: string[] = [];
  lines.push(`export interface ${name} {`);

  if (properties) {
    for (const [propName, propSchema] of Object.entries(properties)) {
      const isRequired = required.has(propName);
      const propType = schemaToTypeString(propSchema, options);
      const quotedName = quotePropertyName(propName);
      lines.push(`  ${quotedName}${isRequired ? "" : "?"}: ${propType};`);
    }
  }

  // Handle additionalProperties
  const additionalProperties = schema["additionalProperties"];
  if (additionalProperties === true) {
    lines.push("  [key: string]: unknown;");
  } else if (
    typeof additionalProperties === "object" &&
    additionalProperties !== null
  ) {
    const additionalType = schemaToTypeString(additionalProperties as AnySchema, options);
    lines.push(`  [key: string]: ${additionalType};`);
  }

  lines.push("}");
  return lines.join("\n");
}

File: packages/python-zod-openapi/src/python_zod_openapi/interface_generator.py
from __future__ import annotations

from .registry import (
    get_schema_exported_variable_name_for_primitive_type,
    get_schema_exported_variable_name_for_string_format,
)
from .types import AnySchema

_VALID_IDENTIFIER = "_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"


def _quote_property_name(name: str) -> str:
    if name and name[0].isalpha() and all(ch in _VALID_IDENTIFIER for ch in name):
        return name
    return repr(name)


def _to_pascal_case(name: str) -> str:
    parts = []
    current = ""
    for ch in name:
        if ch.isalnum():
            current += ch
        else:
            if current:
                parts.append(current)
                current = ""
    if current:
        parts.append(current)
    return "".join(part[:1].upper() + part[1:] for part in parts if part)


def schema_export_name_to_output_alias(name: str) -> str:
    return f"{_to_pascal_case(name)}Output"


def _register_output_schema_name(schema_name: str, output_schema_names: set[str] | None) -> str:
    if output_schema_names is not None:
        output_schema_names.add(schema_name)
    return schema_export_name_to_output_alias(schema_name)


def _get_registered_output_alias(
    schema: AnySchema,
    output_schema_names: set[str] | None = None,
) -> str | None:
    if not isinstance(schema, dict):
        return None

    if schema.get("type") == "string" and isinstance(schema.get("format"), str):
        custom = get_schema_exported_variable_name_for_string_format(schema["format"])
        if custom:
            return _register_output_schema_name(custom, output_schema_names)

    if schema.get("type") in {"number", "integer", "boolean"}:
        custom = get_schema_exported_variable_name_for_primitive_type(schema["type"])
        if custom:
            return _register_output_schema_name(custom, output_schema_names)

    return None


def schema_to_type_string(
    schema: AnySchema,
    *,
    output_schema_names: set[str] | None = None,
) -> str:
    if not isinstance(schema, dict):
        return "Any"

    if isinstance(schema.get("$ref"), str):
        ref = schema["$ref"]
        if ref.startswith("#/components/schemas/"):
            name = ref.split("#/components/schemas/")[1]
            return name.replace("%20", " ")
        return "Any"

    if "oneOf" in schema and isinstance(schema["oneOf"], list):
        members = [
            schema_to_type_string(item, output_schema_names=output_schema_names)
            for item in schema["oneOf"]
        ]
        if not members:
            return "Any"
        return f"Union[{', '.join(members)}]" if len(members) > 1 else members[0]

    if "allOf" in schema and isinstance(schema["allOf"], list):
        members = [
            schema_to_type_string(item, output_schema_names=output_schema_names)
            for item in schema["allOf"]
        ]
        if not members:
            return "Any"
        return f"AllOf[{', '.join(members)}]" if len(members) > 1 else members[0]

    if "anyOf" in schema and isinstance(schema["anyOf"], list):
        members = [
            schema_to_type_string(item, output_schema_names=output_schema_names)
            for item in schema["anyOf"]
        ]
        if not members:
            return "Any"
        return f"Union[{', '.join(members)}]" if len(members) > 1 else members[0]

    schema_type = schema.get("type")

    if schema_type == "string":
        registered = _get_registered_output_alias(schema, output_schema_names)
        if registered:
            return registered
        if isinstance(schema.get("enum"), list):
            values = ", ".join(repr(value) for value in schema["enum"])
            return f"Literal[{values}]"
        return "str"

    if schema_type in {"number", "integer"}:
        registered = _get_registered_output_alias(schema, output_schema_names)
        if registered:
            return registered
        if isinstance(schema.get("enum"), list):
            values = ", ".join(repr(value) for value in schema["enum"])
            return f"Literal[{values}]"
        return "int" if schema_type == "integer" else "float"

    if schema_type == "boolean":
        registered = _get_registered_output_alias(schema, output_schema_names)
        return registered or "bool"

    if schema_type == "null":
        return "None"

    if schema_type == "array":
        items = schema.get("items")
        if isinstance(items, dict):
            item_type = schema_to_type_string(items, output_schema_names=output_schema_names)
            return f"list[{item_type}]"
        return "list[Any]"

    if schema_type == "object" or "properties" in schema:
        return _object_schema_to_type_string(schema, output_schema_names)

    if isinstance(schema.get("enum"), list):
        values = ", ".join(repr(value) for value in schema["enum"])
        return f"Literal[{values}]"

    return "Any"


def _object_schema_to_type_string(
    schema: AnySchema,
    output_schema_names: set[str] | None = None,
) -> str:
    properties = schema.get("properties")
    required = set(schema.get("required", []))
    additional = schema.get("additionalProperties")

    if not properties and not additional:
        return "dict[str, Any]"

    parts: list[str] = []

    if isinstance(properties, dict):
        for prop_name, prop_schema in properties.items():
            prop_type = schema_to_type_string(prop_schema, output_schema_names=output_schema_names)
            quoted = _quote_property_name(prop_name)
            suffix = "" if prop_name in required else "?"
            parts.append(f"{quoted}{suffix}: {prop_type}")

    if additional is True:
        parts.append("[key: str]: Any")
    elif isinstance(additional, dict):
        additional_type = schema_to_type_string(additional, output_schema_names=output_schema_names)
        parts.append(f"[key: str]: {additional_type}")

    return "{ " + "; ".join(parts) + " }"


def generate_interface(
    name: str,
    schema: AnySchema,
    *,
    output_schema_names: set[str] | None = None,
) -> str:
    properties = schema.get("properties") if isinstance(schema, dict) else None

    if not isinstance(schema, dict) or (schema.get("type") != "object" and not properties):
        type_str = schema_to_type_string(schema, output_schema_names=output_schema_names)
        return f"{name}: TypeAlias = {type_str}"

    lines: list[str] = []
    lines.append(f"class {name}(TypedDict):")

    if isinstance(properties, dict):
        required = set(schema.get("required", []))
        for prop_name, prop_schema in properties.items():
            prop_type = schema_to_type_string(prop_schema, output_schema_names=output_schema_names)
            quoted = _quote_property_name(prop_name)
            if prop_name in required:
                lines.append(f"    {quoted}: {prop_type}")
            else:
                lines.append(f"    {quoted}: NotRequired[{prop_type}]")

    additional = schema.get("additionalProperties")
    if additional is True:
        lines.append("    __extra__: dict[str, Any]")
    elif isinstance(additional, dict):
        additional_type = schema_to_type_string(additional, output_schema_names=output_schema_names)
        lines.append(f"    __extra__: dict[str, {additional_type}]")

    if len(lines) == 1:
        lines.append("    pass")

    return "\n".join(lines)
